instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,419 @@
+// not provided in the resource. This function is to filter out the configs that have a real change in their values, it returns
+// a map of these configs and their new values.
+func getChangedConfigMap(d *schema.ResourceData) (map[string]interface{}, error) {
+	newConfig := d.Get("config").([]interface{})[0].(map[string]interface{})
+	changedConfigMap := make(map[string]interface{}, 0)
+	var kafkaTopicConfig KafkaTopicConfigOptions
+	decodedConfigMap := make(map[string]interface{}, 0) // This map stores all non-zero values and all the pointers
+
+	// Decode twice here because zero-value non-pointers need to be removed
+	err := mapstructure.Decode(newConfig, &kafkaTopicConfig)
+	if err != nil {
+		return nil, fmt.Errorf("[Error] Error decoding the config {} block to KafkaTopicConfigOptions: %w", err)
+	}
+	err = mapstructure.Decode(kafkaTopicConfig, &decodedConfigMap)
+	if err != nil {
+		return nil, fmt.Errorf("[Error] Error decoding KafkaTopicConfigOptions to a map: %w", err)
+	}
+
+	// The following block is to get what configs are indeed changed
+	for k, v := range newConfig {
+		key := fmt.Sprintf("config.0.%s", k)
+		valueOfV := reflect.ValueOf(decodedConfigMap[k])
+		if d.HasChange(key) {
+			before, after := d.GetChange(key)
+			isBeforeZeroValue := before == reflect.Zero(reflect.TypeOf(before)).Interface()
+			isAfterZeroValue := after == reflect.Zero(reflect.TypeOf(after)).Interface()
+			if !isBeforeZeroValue && !isAfterZeroValue {
+				changedConfigMap[k] = v
+			}
+			if !isBeforeZeroValue && isAfterZeroValue && (valueOfV.Kind() == reflect.Ptr) {
+				changedConfigMap[k] = v
+			}
+			if isBeforeZeroValue && !isAfterZeroValue {
+				changedConfigMap[k] = v
+			}
+		}
+	}
+	return changedConfigMap, nil
+}
+
+func resourceKafkaTopicDelete(d *schema.ResourceData, meta interface{}) error {
+	cluster_id := d.Get("cluster_id").(string)
+	topic := d.Get("topic").(string)
+	log.Printf("[INFO] Deleting Kafka topic %s in %s.", topic, cluster_id)
+	client := meta.(*Config).Client
+
+	err := client.DeleteKafkaTopic(cluster_id, topic)
+	if err != nil {
+		return fmt.Errorf("[Error] Error deleting Kafka topic %s: %w", topic, err)
+	}
+	d.SetId("")
+	d.Set("cluster_id", "")
+	d.Set("topic", "")
+
+	log.Printf("[INFO] Kafka topic %s has been deleted.", topic)
+	return nil
+}
+
+func resourceKafkaTopicStateImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	idParts := strings.Split(d.Id(), "&")
+	if len(idParts) != 2 || idParts[0] == "" || idParts[1] == "" {
+		return nil, fmt.Errorf("[Error] Unexpected format of ID (%q), expected <CLUSTER-ID>&<TOPIC>", d.Id())
+	}
+	d.Set("cluster_id", idParts[0])
+	d.Set("topic", idParts[1])
+	return []*schema.ResourceData{d}, nil
+}
+
+//customDiffSuppressFunc is used to suppress the diff if a config is not provided in the resource.
+func customDiffSuppressFunc(k, old, new string, d *schema.ResourceData) bool {
```

Comment:
Should have a separate diff method here based on the type of the property. Because it's possible for "0" to be a legitimate input to a string property.

Reply:
Yeah, that's true, will add another function for int.